### ISSUE TITLE
Performance scaling problems

### ISSUE THREAD

**Issue by haltesh** [QuestPDF/QuestPDF#1062]
**Describe the bug**
To preface this, I realize that what I'm about to describe is an extreme example, but when trying to solve it I found multiple issues with the way QuestPDF currently scales.

I have a document with multiple sections, each section containing its own header and a table. Most of the time when composing this document, the number of sections is low, and the document ends up with somewhere around 10-100 pages.

Sometimes, however, this document can contain more than 1000 sections, and end up with over 1000 pages, and when that happens, memory usage spikes (a very large example with 6000 pages can consume close to 7GB for a single document).

To try and combat the high memory usage, I rewrote the document to render everything in a single `DynamicComponent`. This massively reduced the memory usage (7GB -> 300MB), but at a fairly large cost to performance -  the dynamic implementation takes 4 times as long to render. This was a tough pill to swallow, but it still felt better than the alternative - until I tried what happens when rendering multiple documents in parallel.

As it turns out, trying to generate multiple documents at the same time is essentially pointless. When trying to generate 5 documents, generating them sequentially in a single thread is either just as fast, or sometimes even faster than running multiple threads:

```
# Single thread result:

Time elapsed (1): 00:00:12
Time elapsed (2): 00:00:11
Time elapsed (3): 00:00:13
Time elapsed (4): 00:00:11
Time elapsed (5): 00:00:11
Time elapsed: 00:01:01
```

```
# Multiple threads result:

Time elapsed: 00:01:11
```
The memory usage does scale, however, so while running multiple threads does not yield significant benefits, it does cause the memory usage to increase.

**To Reproduce**
I've created a sample application, which is accessible here: https://github.com/haltesh/QuestPDF.PerformanceScaling202411.

**Expected behavior**

In terms of generating a single document, I'd love for either the default implementation to not be as memory hungry, or for the dynamic implementation to not be as slow as it is.

In terms of multithreaded performance, I'd expect running multiple threads to yield at least a moderate benefit.

**Environment**
What version of the library do you use? 2024.10.4
What operating system do you use? (OS type, x64 vs x86 vs arm64): Win11 x64

**Additional context**
I downgraded the library in the sample application to 2023.12.6, the last version that used SkiaSharp (To test this, you'll have to remove the two calls to `.ClampLines()`, which is a newer feature), to see if this is a regression.

The results are very different:
 - Peak memory usage with the largest document is around ~1GB, down from ~3.5GB on 2024.10.4.
 - The performance delta between the regular document and the dynamic re-implementation is much smaller, and the dynamic re-implementation is actually faster than the default implementation with the largest data model.
 - The performance scales much better when generating documents in parallel.

**Comment by MarcinZiabek** [comment]
Thank you for reporting the performance issue. I can confirm that it exists.

As a temporary solution, in the 2024.12.0 release, I have adjusted the maximum concurrency level for QuestPDF. Currently, the library limits usage to two threads. While this is not an ideal solution, it helps reduce processing time and offers improved performance and memory usage in multi-threaded environments.

I have also identified several factors that may contribute to the library not scaling as expected. I plan to investigate these issues further and aim to improve scalability in future releases.

Additionally, to assist with memory usage, I have introduced a preview implementation of the Lazy element. This feature is available in the 2025.1.0-alpha0 release.

Following documentation: 

> Delays the creation of document content and reduces its lifetime, significantly lowering memory usage in large documents containing thousands of pages. This approach also enhances garbage collection efficiency in memory-constrained environments.

> The provided contentBuilder delegate is invoked later in the document generation process. For optimal performance, divide your document into smaller sections, each encapsulated within its own Lazy element. Further optimizations can be achieved by nesting Lazy elements within each other.  However, note that this technique may increase the overall document generation time due to deferred content processing.

I can use the Lazy element like this:

```csharp
Document
.Create(document =>
{
    document.Page(page =>
    {
        page.Content().Column(column =>
        {
            foreach (var i in Enumerable.Range(0, 1000))
            {
                column.Item().Lazy(lazy => lazy.Column(column2 =>
                {
                    foreach (var j in Enumerable.Range(0, 1000))
                    {
                        column2.Item().Text($"{i} x {j}").FontSize(4);
                    }
                }));
            }
        });
    });
})
.GeneratePdf();
```

It decreases memory usage by 90% while reducing generation performance by 33%.

I am eager to investigate whether the memory cost of caching can be reduced. If successful, the library could automatically apply the Lazy element, delivering the best of both worlds: optimal memory efficiency and improved generation performance.

**Comment by haltesh** [comment]
@MarcinZiabek  So I finally got around to testing the 2025.1.0-alpha0 release, and the memory consumption figures when using the `Lazy` element are encouraging. However, the changes to maximum concurrency have me worried about deploying any version after 2024.10.4.

While we do have to generate one of these large documents from time to time, most of the documents we generate are simple, 1-2 page documents. Previously, when we got a request for one of these massive documents, it obviously took very long and consumed a lot of memory - but while that was going on, we could still generate any number of small, single-page documents. The generation times were obviously longer than usual, but still manageable (jumping from ~30 to ~100 or so ms per small document).

With the new maximum concurrency in place,  just two simultaneous requests for a large document that takes ~30 seconds to generate effectively block all other document generation for 30 seconds.

We obviously have work to do on our side, and are looking into running multiple processes, plus delaying generation (a queue system) for large documents, but for right now I think the concurrency limit should be something we can opt out of.

**Comment by MarcinZiabek** [comment]
Thank you for testing.

Could you please profile the generation process? A JetBrains DotTrace snapshot may help me better understand your bottleneck.

Based on my testing, increasing the level of parallelism above two was not optimal. It not only reduced performance but also increased memory usage.

Of course, it is only a temporary solution. I already started investigating the root cause of reduced scalability. The problem is likely complex and there are several factors. I would like to introduce regular enhancements to improve the matter.

**Comment by MarcinZiabek** [comment]
With the 2025.1 release, several improvements have been introduced to enhance overall performance while significantly reducing memory consumption and garbage collector pressure.

Additionally, for large documents, a dedicated Lazy API is now available to further optimize resource usage. Thanks to these enhancements, we’ve also increased the maximum concurrency level to 4.

Since there has been no recent activity, I’m closing this issue. If you have a similar request in the future, please feel free to open a new GitHub issue. Thanks for your feedback! ❤️

**Comment by chuongxl** [comment]
I upgrade from 2022 to 2025 and the creation time increase from 21s to 50 second for 50 pages.
It’s very bad for my application.
WE expect the creation time bellow  20s. 
The more bad things is we have just bought a licensee. 🥵
